dvc/scm.py

```diff
@@ -123,3 +126,61 @@ def resolve_rev(scm: "Git", rev: str) -> str:
             if len(ref_infos) > 1:
                 raise RevError(f"ambiguous Git revision '{rev}'")
         raise RevError(str(exc))
+
+
+def get_revs_from_args(
+    scm: "Git",
+    head_revs: Optional[List[str]] = None,
+    num: int = 1,
+    all_branches: bool = False,
+    all_tags: bool = False,
+    all_commits: bool = False,
+) -> OrderedDict:
+    from dvc.repo.experiments.utils import fix_exp_head
+
+    if num < 1 and num != -1:
+        raise InvalidArgumentError(f"Invalid number of commits '{num}'")
+
+    if not any([head_revs, all_branches, all_tags, all_commits]):
+        return OrderedDict()
+
+    head_revs = head_revs or []
+    revs = []
+    for rev in head_revs:
+        n = 0
+        while True:
+            if num == n:
+                break
+            try:
+                head = fix_exp_head(scm, f"{rev}~{n}")
+                assert head
+                revs.append(resolve_rev(scm, head))
+            except RevError:
+                break
+            n += 1
+
+    revs = get_revs_from_flags(scm, revs, all_branches, all_tags, all_commits)
+    rev_resolver = partial(resolve_rev, scm)
+    result = OrderedDict(
+        (sha, None) for sha, _ in group_by(rev_resolver, revs).items()
+    )
+    return result
+
+
+def get_revs_from_flags(
+    scm: "Git",
+    revs: List[str],
+    all_branches: bool = False,
+    all_tags: bool = False,
+    all_commits: bool = False,
+) -> List[str]:
+    return_revs = revs.copy()
+    if all_commits:
+        return_revs.extend(scm.list_all_commits())
+    else:
+        if all_branches:
+            return_revs.extend(scm.list_branches())
+
+        if all_tags:
+            return_revs.extend(scm.list_tags())
+    return return_revs
```

Comment:
This can just be `iter_revs` as described by @skshetry (or maybe something like `iter_rev_names`), and it can yield the same `sha, names` pairs that brancher currently uses. It should also support `all_experiments` the same way that brancher does. So basically this iterator should do everything that `brancher` does right now except for modifying `repo` and generating a git-fs.

So this method would look something like
```py
def iter_rev_names(
    scm: "Git",
    revs: List[str],
    all_branches: bool = False,
    ...
) Generator[...]:
    ... # get revs list the same way as brancher()
    if revs:
            rev_resolver = partial(resolve_rev, scm)
            # yield (rev, names) pairs
            yield from group_by(rev_resolver, revs).items()
```
This way code outside of brancher can use the `rev, names` items as needed without repo modification and git-fs overhead.

And now `brancher()` itself could look something like
```py
def brancher(
    self,
    *args,
    sha_only=False,
    **kwargs,
):
    saved_fs = self.fs
    try:
        for sha, names in iter_rev_names(self.scm, *args, **kwargs):
                self.fs = GitFileSystem(scm=scm, rev=sha)
                if self.fs.exists(self.root_dir):
                    if sha_only:
                        yield sha
                    else:
                        yield ", ".join(names)
```